File: MiniGame-NET/MiniGameEngine/AssetsLoader.cs
// Copyright 2023 Peter Bakota
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

using MiniGameEngine.Utils;
using SDL2;

namespace MiniGameEngine
{
    public abstract class AssetsLoader
    {
        #pragma warning disable 8618
        internal static string _assetsFolder;
        internal static readonly string _cwd = Environment.CurrentDirectory;

        protected AssetsLoader(string assetsFolder)
        {
            _assetsFolder = assetsFolder;
        }

        protected static IntPtr LoadBitmap(IntPtr renderer, string path)
        {
            var fullpath = Path.Combine(_cwd, _assetsFolder, path);

            var surface = SDLUtil.CheckSDLErr(() => SDL_image.IMG_Load(fullpath));
            var texture = SDLUtil.CheckSDLErr(() => SDL.SDL_CreateTextureFromSurface(renderer, surface));

            return texture;
        }
        
        protected static Sound.Audio LoadSoundFx(string path)
        {
            var fullpath = Path.Combine(_cwd, _assetsFolder, path);

            var audio = SDL.SDL_LoadWAV(fullpath, out var specs, out var buffer, out var len);
            if(audio == IntPtr.Zero) {
                SDLUtil.LogErr($"Unable to read audio wav file {fullpath}, Error:{SDL.SDL_GetError()}");
                Environment.Exit(1);
            }

            return new Sound.Audio
            {
                buffer = buffer,
                length = len,
                audio = specs,
            };
        }

        public static string GetAssetPath(string fileName)
            => Path.Combine(_cwd, _assetsFolder, fileName);
    }
}

File: MiniGame-Java/src/main/java/rs/lab/mges/engine/AssetsLoader.java
package rs.lab.mges.engine;

import io.github.libsdl4j.api.render.SDL_Renderer;
import io.github.libsdl4j.api.render.SDL_Texture;
import io.github.libsdl4j.api.render.SdlRender;
import io.github.libsdl4j.api.surface.SDL_Surface;
import io.github.libsdl4j.api.surface.SdlSurface;
import java.nio.file.Path;

import static io.github.libsdl4j.api.error.SdlError.SDL_GetError;

public abstract class AssetsLoader {

    public static final String CWD = Helpers.getHere().toString();
    public static String assetsFolder;

    public static void setAssetsFolder(String sf) {
        assetsFolder = sf;
    }

    public static SDL_Texture loadBitmap(SDL_Renderer renderer, String path) {
        final var fullPath = Path.of(CWD, assetsFolder, path);
        var surface = (SDL_Surface) SDLUtils.CheckSDLErrPointer(() -> SdlSurface.SDL_LoadBMP(fullPath.toString()));
        var texture = (SDL_Texture) SDLUtils.CheckSDLErrPointer(() -> SdlRender.SDL_CreateTextureFromSurface(renderer, surface));

        SdlSurface.SDL_FreeSurface(surface);
        return texture;
    }

    public static SDL_Surface loadBitmapFont(String filename) {
        var fullPath = getAssetPath(filename);
        var surface = (SDL_Surface) SDLUtils.CheckSDLErrPointer(() -> SdlSurface.SDL_LoadBMP(fullPath));
        return surface;
    }
    
    public static Sound.Audio loadSoundFx(String path) {
        var fullpath = getAssetPath(path);
        var newAudio = Sound.createAudio(fullpath, false, 100);
        if (newAudio == null) {
            SDLUtils.LogErr("Unable to read audio wav file %s, Error: %s", fullpath, SDL_GetError());
            System.exit(1);
        }
        return newAudio;
    }

    public static String getAssetPath(String fileName) {
        return Path.of(CWD, assetsFolder, fileName).toString();
    }
}
